### ISSUE TITLE
reproducible HTTP 502 error in OAuth /oauth/authorize

### ISSUE THREAD

**Issue by snarfed** [tootsuite/mastodon#12915]
Hi all! I keep hitting a 502 in `/oauth/authorize` in 3.0.1. Is this a known issue? Here are the steps to reproduce:

1. Create an app, eg:
    ```
    POST https://mastodon.technology/api/v1/apps
    client_name=Bridgy&redirect_uris=http://localhost:8080/mastodon/delete/finish%0Ahttp://localhost:8080/delete/finish%0Ahttp://localhost:8080/publish/mastodon/finish%0Ahttp://localhost:8080/mastodon/callback&website=http://localhost:8080&scopes=read+read:accounts+read:blocks+read:favourites+read:filters+read:follows+read:lists+read:mutes+read:notifications+read:reports+read:search+read:statuses+write+write:accounts+write:blocks+write:favourites+write:filters+write:follows+write:lists+write:media+write:mutes+write:notifications+write:reports+write:statuses+follow+push
    ```
    (I've URL-unescaped everything except the newlines, here and below, for readability.)
1. Authorize with an initial set of read only scopes, eg:
    ```
    https://mastodon.technology/oauth/authorize?response_type=code&client_id=...&client_secret=...&scope=read:accounts read:accounts read:blocks read:search&redirect_uri=http%3A%2F%2Flocalhost%3A8080%2Fmastodon%2Fcallback&state={"app_key":"...","state":"%257B%2522feature%2522%253A%2522listen%2522%252C%2522operation%2522%253A%2522add%2522%257D"}
    ```
    ...and get an access token with `/oauth/token`.
1. Authorize with an expanded set of scopes, including writes, eg:
    ```
    https://mastodon.technology/oauth/authorize?response_type=code&client_id=...&client_secret=...&scope=read:accounts read:accounts read:blocks read:search write:statuses write:favourites write:media&redirect_uri=http://localhost:8080/mastodon/callback&state={"app_key":"...","state":"%7B%22feature%22%3A%22publish%22%2C%22operation%22%3A%22add%22%7D"}
    ```
    ...and get an access token with `/oauth/token`.
1. Authorize with the same expanded set of scopes and a longer `state`, eg:
    ```
    https://mastodon.technology/oauth/authorize?response_type=code&client_id=...&client_secret=...&scope=read:accounts read:accounts read:blocks read:search write:statuses write:favourites write:media&redirect_uri=http://localhost:8080/publish/mastodon/finish&state={"app_key":"...","state":"%7B%22include_link%22%3A%22omit%22%2C%22source_key%22%3A%22...%22%2C%22source_url%22%3A%22http%3A%2F%2Flocalhost%2Fpost3.html%22%2C%22target_url%22%3A%22https%3A%2F%2Fbrid.gy%2Fpublish%2Fmastodon%22%7D"}
    ```

This last request 502s with the standard _We're sorry, but something went wrong on our end._ error page.

I tried three 3.0.1 instances just now - 2020-01-21 05:36:12 UTC on mastodon.technology, 05:49:03 mas.to, 06:10:37 mstdn.social - and they all behaved the same way.

Any idea what's going on? My only guess is that the `state` in the last request is longer than some limit, but that seems unlikely. (It was originally 331 chars before I redacted parts of it.)

@ashfurrow, any chance you could look through the mastodon.technology logs around that time ^ and see if you can correlate the `/oauth/authorize` 502 with anything?

Thanks in advance!

**Comment by ashfurrow** [comment]
I took some time looking through the logs but couldn't find anything helpful, sorry!

**Comment by snarfed** [comment]
My users are continuing to hit this on a growing number of interactions with Mastodon OAuth, [now sometimes even on initial login/auth.](https://chat.indieweb.org/dev/2022-01-04#t1641257142314800) I may look at debugging and fixing it myself soon.

**Comment by snarfed** [comment]
Haven't yet been able to reproduce this against a local Mastodon server built from HEAD. It handled this 785-character URL fine:

```
http://localhost:3000/oauth/authorize?response_type=code&client_id=2DgqdtD0K9EU2_o3RFEOdLwr6UwoL_92ZmrQCl4sB1g&client_secret=ooKrV-BCFgc1gWL4RaySw5o8GQH3XH9H72EMpr5eJ1U&scope=read:accounts%20read:blocks%20read:bookmarks%20read:favourites%20read:filters%20read:follows%20read:lists%20read:mutes%20read:notifications%20read:search%20read:statuses%20write%20write:accounts%20write:blocks%20write:bookmarks%20write:favourites%20write:filters%20write:follows%20write:lists%20write:media%20write:mutes%20write:notifications%20write:reports%20write:statuses%20follow%20push&redirect_uri=http%3A%2F%2Flocalhost%3A8080%2Fmastodon%2Fcallback&state={"app_key":"agNhcHByEQsSC01hc3RvZG9uQXBwGAEM","state":"%257B%2522feature%2522%253A%2522listen%2522%252C%2522operation%2522%253A%2522add%2522%257D"}
```

**Comment by snarfed** [comment]
Production Mastodon instances still 502 though, eg mastodon.technology:

```
https://mastodon.technology/oauth/authorize?response_type=code&client_id=...&client_secret=...&scope=read:accounts%20read:blocks%20read:notifications%20read:search%20read:statuses&redirect_uri=https%3A%2F%2Fbrid.gy%2Fmastodon%2Fcallback&state=%7B%22app_key%22:%22agdicmlkLWd5chgLEgtNYXN0b2RvbkFwcBiAgKjUhuuPCgw%22,%22state%22:%22%257B%2522feature%2522%253A%2522listen%2522%252C%2522operation%2522%253A%2522add%2522%257D%22%7D
```

(Not that exact URL, of course, since I've redacted the client id and secret. Happy to provide the exact URL somewhere non-public if anyone wants to dig it.)

**Comment by metbril** [comment]
I have a similar issue since this morning with mastodon.social running 3.4.6.

**Comment by Gargron** [comment]
I suspect that it doesn't happen on localhost because the error is returned by nginx. Mastodon does not return HTTP 502 errors on its own. HTTP 502 means "Bad gateway" and means nginx could not get or process the response from the gateway (Mastodon). 

I believe I found the corresponding error in /var/log/nginx/error.log:

> upstream sent too big header while reading response header from upstream

**Comment by Gargron** [comment]
The "solution" is to increase the size of the proxy buffers in nginx, but this isn't really a distributable solution, and I'm concerned about the implications on other types of requests. It seems like it could be a mistake on behalf of the [doorkeeper ](https://github.com/doorkeeper-gem/doorkeeper) gem in sending back a header that's way too long.

**Comment by snarfed** [comment]
Thank you for investigating!

**Comment by mlncn** [comment]
Any mitigations identified (even to go from very flaky to somewhat flaky) or possible for the workaround nginx proxy buffer increase to be done for mastodon.social ?  Currently blocking all [signup attempts from Bridgy](https://brid.gy/mastodon/start?x=69&y=17&feature=listen&instance=&scope=read%3Aaccounts+read%3Ablocks+read%3Anotifications+read%3Asearch+read%3Astatuses)

**Comment by Nezteb** [comment]
Can confirm @mlncn's workaround:
1. Go to https://mastodon.social/
2. Sign out of Mastodon
3. Go to https://brid.gy/
4. Click on the Mastodon button
5. Sign into Mastodon
6. Enjoy your sweet sweet Bridgy/Mastodon connection

**Comment by snarfed** [comment]
@Nezteb @mlncn confirmed, that workaround works for me too! I'll follow up in https://github.com/doorkeeper-gem/doorkeeper/issues/1554

**Comment by thiagomgd** [comment]
workaround was working for me, but stopped over the weekend. Tested on 2 computers (both macOs, chrome), on regular window and also incognito

**Comment by timnolte** [comment]
> The "solution" is to increase the size of the proxy buffers in nginx, but this isn't really a distributable solution, and I'm concerned about the implications on other types of requests. It seems like it could be a mistake on behalf of the [doorkeeper ](https://github.com/doorkeeper-gem/doorkeeper) gem in sending back a header that's way too long.

I can confirm that using the proxy buffer settings mentioned [here](https://github.com/doorkeeper-gem/doorkeeper/issues/1554#issuecomment-1304606633) for my Mastodon instance Nginx configuration did indeed fix the 502 errors.

**Comment by snarfed** [comment]
My (and other) users keep hitting this and getting confused and blocked by it. It's preventing a sizeable fraction of them from using my service at all.

@Gargron @ClearlyClaire and other maintainers, I understand that you're reluctant to increase the nginx proxy buffer sizes, but as people have confirmed here and in https://github.com/doorkeeper-gem/doorkeeper/issues/1554#issuecomment-1304606633, it clearly does fix this problem. The lengths of the redirect URLs and headers from doorkeeper seem correct, not a mistake. Could we increase the proxy buffers a moderate amount, so that it's hopefully not too invasive but still solves the problem?

**Comment by snarfed** [comment]
I ended up working around this by storing the OAuth redirect state in memory on my end, which let me reduce the redirect URL length. I'll leave this open to track on Mastodon's end.

**Comment by 33b5e5** [comment]
> I can confirm that using the proxy buffer settings mentioned [here](https://github.com/doorkeeper-gem/doorkeeper/issues/1554#issuecomment-1304606633) for my Mastodon instance Nginx configuration did indeed fix the 502 errors.

Thank you. I added these three lines:

```
proxy_buffers 4 16k;
proxy_buffer_size 16k;
proxy_busy_buffers_size 32k;
```

I think they could technically go in a http, server, or location block, but I chose to add them to the `location @proxy` block that proxies requests to `http://backend`. My thinking is this will cover the endpoints exhibiting the problem, while limiting the impact of the change otherwise.

**Comment by ThisIsMissEm** [comment]
I've asked about this on the OAuth mailing list: https://mailarchive.ietf.org/arch/msg/oauth/VnHry5n5-58Qikk_x8VibrBe55A/

**Comment by ThisIsMissEm** [comment]
@snarfed I'm not sure if this is still an issue for bridgy, but generally the `state` parameter isn't intended for storing arbitrary amounts of data with the oauth request flow, but instead for storing an identifier for that state (and the application should store any necessary data keyed by that identifier), it was more intended as a security feature rather than a "pass arbitrary data through the authorization flow" mechanism

**Comment by snarfed** [comment]
Thanks @ThisIsMissEm! I switched to putting just a identifier in `state` a couple years ago, https://github.com/mastodon/mastodon/issues/12915#issuecomment-1378217764, like you mention.

Interesting to hear that shift in expectation for `state` in OAuth! I definitely believe you that it was never supposed to hold a lot of data, but I'm pretty sure it was originally intended broadly for arbitrary client state, not just security-relevant state, at least going back to OAuth 1 and 1.1. I don't remember that changing in OAuth2, at least initially, but I may not have been paying close enough attention.

**Comment by ThisIsMissEm** [comment]
I guess we could limit the size of state to like 8k minus the average size of the OAuth redirect response. Then we'd just dead-end in pre-auth "state parameter too large"

Or we could edit the dist/nginx.conf to set
``` 
proxy_buffers 4 16k;
proxy_buffer_size 16k;
proxy_busy_buffers_size 32k;
```
As above, but even then there's a limit

**Comment by ThisIsMissEm** [comment]
I think this is probably a won't fix, or if we do make a change, it'd be to limit the state parameter size, even though that would be non-standard. It would be interesting to see if this works by using a POST request for the authorization endpoint? Though you'd still have `state` in the redirect URL, so `Location` header, unless we use `response_mode=form_post` (though I've rarely seen that used)

https://mailarchive.ietf.org/arch/msg/oauth/4SRg0nSsfnzGQkTZIsoBQoqPI6k/
